mutation_reader.cc

```diff
@@ -1156,29 +1178,57 @@ future<> shard_reader::remote_reader::do_fill_buffer(flat_mutation_reader& reade
     });
 }
 
-future<> shard_reader::remote_reader::ensure_buffer_contains_all_fragments_for_last_pos(flat_mutation_reader& reader,
-        circular_buffer<mutation_fragment>& buffer, db::timeout_clock::time_point timeout) {
-    if (buffer.empty() || !buffer.back().is_range_tombstone()) {
-        return make_ready_future<>();
-    }
-
-    auto stop = [this, &reader, &buffer] {
-        if (reader.is_buffer_empty()) {
-            return reader.is_end_of_stream();
-        }
-        const auto& next_pos = reader.peek_buffer().position();
-        if (next_pos.region() != partition_region::clustered) {
-            return true;
-        }
-        return !next_pos.key().equal(*_schema, buffer.back().position().key());
-    };
-
-    return do_until(stop, [this, &reader, &buffer, timeout] {
+future<> shard_reader::remote_reader::fill_buffer(flat_mutation_reader& reader, circular_buffer<mutation_fragment>& buffer,
+        db::timeout_clock::time_point timeout) {
+    return do_fill_buffer(reader, timeout).then([this, &reader, &buffer, timeout] {
         if (reader.is_buffer_empty()) {
-            return do_fill_buffer(reader, timeout);
+            return make_ready_future<>();
         }
-        buffer.emplace_back(reader.pop_mutation_fragment());
-        return make_ready_future<>();
+        buffer = reader.detach_buffer();
+        auto stop = [this, &reader, &buffer] {
+            // The only problematic fragment kind is the range tombstone.
+            // All other fragment kinds are safe to end the buffer on, and
+            // are guaranteed to represent progress vs. the last buffer fill.
+            if (!buffer.back().is_range_tombstone()) {
+                return true;
+            }
+            if (reader.is_buffer_empty()) {
+                return reader.is_end_of_stream();
+            }
+            const auto& next_pos = reader.peek_buffer().position();
+            // To ensure safe progress we have to ensure the following:
+            //
+            // _next_position_in_partition < buffer.back().position() < next_pos
+            //
+            // * The first condition is to ensure we made progress since the
+            // last buffer fill. Otherwise we might get into an endless loop if
+            // the reader is recreated after each `fill_buffer()` call.
```

Comment:
Isn't `_next_position_in_partition`, at the time of recreation, guaranteed to be always greater than `buffer.back().position`? 

We ensure that by the second check: `_tri_cmp(buffer.back().position(), next_pos) < 0`

Reply:
Can you give an example sequence of events which leads to an endless loop without the first check?

Reply:
Is it because we may emit fragments which start before the requested ranges start?

Reply:
It sorta is. But not necessarily. Consider the following:
```
read_buffer={..., {rt, pos=[1, 100)}, {cr, pos=1}}
```
When we recreate the reader with pos `(1, +inf)` clustering range, the underlying reader might re-emit all those range tombstones with `pos=[1, 100)`, with the position trimmed to the clustering range: `pos=(1, 100)`. If we again stop at this position we will be unable to make progress if the reader is evicted after each fill buffer call. This scenario happened in the tests.

Reply:
On second though, this might not be the case now that we use the next position. Let me try.

Reply:
I removed this condition and the test hangs.

Reply:
I would like to understand what exactly happens.

Reply:
Consider the following stream:
`f0=rt{k1, -1}, f1=cr{k1, 0}`
If `f0` is the last element in the buffer, we recreate the reader with `[k1, +inf)` clustering range, however the underlying reader will re-emit the range tombstone with `rt{k1, -1}`, which will again become the last element in the buffer (buffer_size=1). We have an infinite loop. This is a contrived test scenario which stresses the reader recreation as much as possible, however a situation like this can theoretically happen in a real server, given enough range tombstones and a sufficient amount of bad luck.